internal/overlord/checkstate/manager.go

```diff
@@ -0,0 +1,216 @@
+			continue
+		}
+		if len(names) > 0 && !containsString(names, check.config.Name) {
+			continue
+		}
+		infos = append(infos, check.info())
+	}
+
+	sort.Slice(infos, func(i, j int) bool {
+		return infos[i].Name < infos[j].Name
+	})
+	return infos, nil
+}
+
+func containsString(slice []string, value string) bool {
+	for _, v := range slice {
+		if v == value {
+			return true
+		}
+	}
+	return false
+}
+
+// CheckInfo provides status information about a single check.
+type CheckInfo struct {
+	Name         string
+	Healthy      bool
+	Failures     int
+	LastError    string
+	ErrorDetails string
+}
+
+// checkData holds state for an active health check.
+type checkData struct {
+	config  *plan.Check
+	checker checker
+	done    chan struct{}
+
+	mutex     sync.Mutex
+	failures  int
+	action    FailureFunc
+	actionRan bool
+	lastErr   error
+	cancel    context.CancelFunc
+}
+
+type checker interface {
+	check(ctx context.Context) error
+}
+
+func (c *checkData) loop() {
+	logger.Debugf("Check %q starting with period %v", c.config.Name, c.config.Period.Value)
+	defer logger.Debugf("Check %q stopped", c.config.Name)
+
+	ticker := time.NewTicker(c.config.Period.Value)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.check()
+		case <-c.done:
+			return
+		}
+	}
+}
+
+func (c *checkData) check() {
+	// Set up context: a timeout or call to c.stop() will cancel the check.
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.Timeout.Value)
+	defer cancel()
+	c.mutex.Lock()
+	c.cancel = cancel
+	c.mutex.Unlock()
+
+	// Run the check!
+	err := c.checker.check(ctx)
+
+	// Lock while we update state, as the manager may access these too.
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.cancel = nil
+	c.lastErr = err
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			// Check was stopped, don't trigger failure action.
+			logger.Debugf("Check %q canceled in flight", c.config.Name)
+			return
+		}
+
+		// Track failure, run failure action if "failures" threshold was hit.
+		c.failures++
+		logger.Noticef("Check %q failure %d (threshold %d): %v",
+			c.config.Name, c.failures, c.config.Failures, err)
+		if !c.actionRan && c.failures >= c.config.Failures {
+			logger.Noticef("Check %q failure threshold %d hit, triggering action",
+				c.config.Name, c.config.Failures)
+			c.action(c.config.Name)
+			c.actionRan = true
+		}
+		return
+	}
+	c.failures = 0
+	c.actionRan = false
+}
+
+func (c *checkData) stop() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.cancel != nil {
+		c.cancel()
+	}
+	close(c.done)
+}
+
+func (c *checkData) info() *CheckInfo {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	info := &CheckInfo{
+		Name:     c.config.Name,
+		Healthy:  c.failures == 0,
+		Failures: c.failures,
+	}
+	if c.lastErr != nil {
+		info.LastError = c.lastErr.Error()
+		switch e := c.lastErr.(type) {
+		case *outputError:
+			info.ErrorDetails = e.output()
+		}
```

Comment:
Other errors just ignored?

Reply:
No, other errors are stored in `info.LastError` (the `err.Error()` string). It's just errors that have additional detail that we'll include here -- only the exec check's `outputError` for now.